File: ml-benchmark/synduce/list/hamming.f
Inductive ZipCList = cnil Unit| single {Int, Int} | concat {ZipCList, ZipCList};
Inductive ZipList = nil Unit | cons {Int, Int, ZipList};

cat = fix (
  \f: ZipList -> ZipList -> ZipList. \xs: ZipList. \ys: ZipList.
  match xs with
    nil _ -> ys
  | cons {a, b, t} -> cons {a, b, (f t ys)}
  end
);

repr = fix (
  \f: ZipCList -> Compress ZipList. \xs: ZipCList.
  match xs with
    cnil _ -> nil unit
  | single {a, b} -> cons {a, b, nil unit}
  | concat {a, b} -> cat (f a) (f b)
  end
);

spec = fix (
  \f: ZipList -> Int. \xs: ZipList.
  match xs with
    nil _ -> 0
  | cons {a, b, t} -> if (== a b) then + 1 (f t) else (f t)
  end
);

main = \xs: ZipCList. spec (repr xs);

File: ml-benchmark/synduce/list/hamming.ml
type zipCList = Cnil of unit | Single of int * int | Concat of zipCList * zipCList
type zipList = Nil of unit | Cons of int * int * zipList

let rec cat xs ys =
  match xs with
  | Nil _ -> ys
  | Cons (a, b, t) -> Cons (a, b, cat t ys)

val repr: zipCList -> zipList compress
let rec repr xs =
  match xs with
  | Cnil _ -> Nil ()
  | Single (a, b) -> Cons (a, b, Nil ())
  | Concat (a, b) -> cat (repr a) (repr b)

let rec spec xs =
  match xs with
  | Nil _ -> 0
  | Cons (a, b, t) -> if a == b then 1 + spec t else spec t

let program xs = spec (repr xs)